### PR TITLE
Feature: Router 1

### PR DESCRIPTION
**Pull request by YehorPytomets**
### Changes introduced by this PR:
* Add `Router` and `routing` behaviour. This allows to switch content of page without requesting server.
* Add `NotFoundPage`.

### CODE REVIEW

File: webclient/app/router.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    this._beforeRenderPageHandlers.push(handler);
+  }
+
+  /**
+   * Renders ui page, mapped to specified hash.
+   * Before rendering new page, clears content root container, e.g. previous page.
+   *
+   * @param {string} hash - is a hash, which may be mapped to any ui components.
+   */
+  renderPage(hash) {
```
Comment (dkalinin) [inline webclient/app/router.js:87]: Should be private.

File: webclient/app/router.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  /**
+   * Retrieves page for provided hash or 'not-found' otherwise.
+   *
+   * @param {string} hash - hash value. In '#/val' hash value is '/val'.
+   * @return {function(): Component} page to render.
+   */
+  _retrievePageCreator(hash) {
+    return this._pageMappings[hash] ? this._pageMappings[hash] : this._properties.notFoundPageCreator;
```
Comment (dkalinin) [inline webclient/app/router.js:99]: Move ` this._pageMappings[hash]` to variable.

File: webclient/app/router.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    this._hashChangeHandlers.push((hash) => {
+      this._beforeRenderPageHandlers.forEach((handler) => handler(hash));
+      this.renderPage(hash);
+    });
+
+    _window.addEventListener('hashchange', () => {
+      const hash = this._getUrlHash();
+      this._hashChangeHandlers.forEach((handler) => handler(hash));
+    });
```
Comment (dkalinin) [inline webclient/app/router.js:50]: Please move this to a separate method.

File: webclient/app/router.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    _window.addEventListener('load', () => {
```
Comment (dkalinin) [inline webclient/app/router.js:52]: Load is not needed.

File: webclient/tests/app/router-tests.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        [loginHash]: () => new LoginForm(fixture),
+      };
+    },
+  });
+
+  test("should have default hash.", (assert) => {
+    assert.expect(1);
+
+    const expectedDefaultHash = '#/login';
+    const windowMock = {
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:23]: new WindowMock()

File: webclient/tests/app/router-tests.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      },
+      addEventListener(event, listener) {
+      }
+    };
+    const router = new Router(windowMock, null, {
+      defaultUrlHash: expectedDefaultHash,
+      notFoundPageCreator: null
+    });
+
+    const actualDefaultHash = router.defaultUrlHash;
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:35]:
```suggestion
    const actualDefaultHash = windowMock.location.hash;
```

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review

File: webclient/app/router.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    _window.addEventListener('hashchange', () => {
+      const hash = this._getUrlHash();
+      this._executeHashChangedHandlers(hash);
+    });
+
+    this._proceedInitialRendering();
+  }
+
+  _proceedInitialRendering() {
+    const {defaultUrlHash} = this._properties;
```
Comment (dkalinin) [inline webclient/app/router.js:56]: - [ ] Please move this line to else block.

File: webclient/app/router.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+    this._hashChangeHandlers.push((hash) => {
+      this._beforeRenderPageHandlers.forEach((handler) => handler(hash));
+      this._renderPage(hash);
+    });
+
+    _window.addEventListener('hashchange', () => {
+      const hash = this._getUrlHash();
+      this._executeHashChangedHandlers(hash);
+    });
```
Comment (dkalinin) [inline webclient/app/router.js:50]: - [ ] This doesn't handle empty hash. Just use the same logic as in `_proceedInitialRendering ` and refactor.

File: webclient/tests/app/router-tests.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    });
+
+    const expectedHash = `#${routerDefaultHash}`;
+    const actualHash = windowMock.location.hash;
+    assert.equal(actualHash, expectedHash, 'Should properly set default hash to a window object.');
+
+    assert.verifySteps(['Should trigger default hash handler.'], 'Default hash handler was triggered.');
+  });
+
+  test("should render default page.", (assert) => {
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:32]: - [ ] This case is already verified in previous test.

File: webclient/tests/app/router-tests.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    const defaultUrlHash = '/login';
+    const notExistingUrlHash = '/not-existing';
+    const windowMock = new WindowMock();
+    const pageMapping = {
+      [defaultUrlHash]: () => {
+        assert.notOk(true, 'Should not page creator from pageMapping.');
+      },
+    };
+    new Router(windowMock, pageMapping, {
+      defaultUrlHash: notExistingUrlHash,
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:62]: - [ ] This doesn't make sense.

File: webclient/tests/app/router-tests.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+      notFoundPageCreator: () => {
+        assert.step('Should call \'not-found-page\' creator.');
+      },
+    });
+    windowMock.location.hash = `#${notExistingUrlHash}`;
+
+    assert.verifySteps([
+      'Should call \'not-found-page\' creator.',
+      'Should call \'not-found-page\' creator.',
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:71]: 'not-found-page' should be 

- [ ] called once.

File: webclient/tests/app/router-tests.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    const router = new Router(windowMock, pageMapMock, {
+      defaultUrlHash: loginHash,
+      notFoundPageCreator: null,
+    });
+    router.beforeRenderPage(() => {
+      assert.step('Should trigger handler.');
+    });
+    windowMock.location.hash = `#${loginHash}`;
+
+    assert.verifySteps(['Should trigger handler.'], 'Handler was triggered.');
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:92]: - [ ] I think that `beforeRenderPage` should be called twice.

File: webclient/tests/app/router-tests.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+  });
+})();
+
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:95]: Please add test case with the following scenario:
1. Perform initial rendering. -> Default page is rendered.
2. Change hash to empty string. -> Default page is rendered again.

- [ ] Done.

File: webclient/tests/app/router-tests.js (hunk 12 above)
Comment (YehorPytomets) [reply webclient/tests/app/router-tests.js:92]: Agree on that.
Main idea of `beforeRenderPage` is to clear html content of `rootElement` somewhere outside of `router` in order to maintain decoupling.
But there also a problem: `beforeRenderPage` handler get registered only after initial rendering was proceeded, what is wrong.
In order to solve the problem I've decided to pass `rootElement` through constructor and clear it's content inside `router` each time a new page is about to render. And also, relative test was rewritten.

I've thought about another way, like adding `beforeRenderPage` handler in constructor, but I'm not sure that this way is correct.

**Comment by YehorPytomets** [comment]
@dkalinin , PR is ready to review.

File: webclient/app/router.js [hunk 14]
```diff
@@ -1,68 +1,94 @@
...
+    this._proceedPageRendering();
   }
 
-  _proceedInitialRendering() {
-    const {defaultUrlHash} = this._properties;
+  /**
+   * Proceeds page rendering, considering current url.
+   * @private
+   */
+  _proceedPageRendering() {
```
Comment (dkalinin) [inline webclient/app/router.js:76]: - [ ] Please rename according to what is done.

File: webclient/tests/app/router-tests.js [hunk 15]
```diff
@@ -29,67 +29,63 @@ export default (() => {
...
+    assert.verifySteps([
+      'Should clear inner html of root element. Content: .',
+      'Should call default page creator.',
+      'Should clear inner html of root element. Content: .',
+      'Should call default page creator.',
+      'Should clear inner html of root element. Content: .',
+      'Should call default page creator.',
+    ], 'Should clear inner html of root element each time page is rendered.');
   });
 })();
```
Comment (dkalinin) [inline webclient/tests/app/router-tests.js:90]: - [ ] Please also add test case when window.hash is already defined when router is being created.

**Pull request merged**